### ISSUE TITLE
Patroni does not failover to replica when leader postgresql service fails

### ISSUE THREAD

**Issue by thanasis-liapis** [zalando/patroni#1826]
Hello here,

I would like to describe a situation and ask for an explanation if possible. 

I've been doing some tests on creating a HA postgresql cluster with Patroni. The first tests were stopping, restarting, killing patroni service and the result was the expected, to failover to one of the replicas. But when I kill postgresql and simulate a case where the service won't be able to recover, patroni does not failover to a replica, even after the master_start_timeout is reached.

According to the description on the documentation, if I understand correctly, the expected behavior would be to switch to a replica after the timeout of 300 seconds is reached. 

Instead, I get a recurring try of patroni to start postgres on the leader and the following error to the replica...

```    
INFO: Got response from tsdb1 http://192.168.255.125:8008/patroni: {"state": "unknown", "role": "master", "database_system_identifier": "6919798977095709312", "patroni": {"version": "2.0.1", "scope": "clearview"}}
WARNING: Master (tsdb1) is still alive
INFO: following a different leader because i am not the healthiest node
```

Could you please advise? Is this the expected behavior? How can I trigger a failover procedure in such a case?

Thanks in advance

**Comment by CyberDem0n** [comment]
> when I kill postgresql and simulate a case where the service won't be able to recover

What exactly have you done?

What is in Patroni logs on all nodes?

How does the Patroni config look like?

And please pay attention, that I am asking all these questions only because you didn't follow the template.

**Comment by thanasis-liapis** [comment]
Thanks for your quick answer and I apologise for not including this info in my first message!

The use-case is the following. I have created a two-node postgresql cluster with patroni on top. What I am testing is to make the postgresql on the leader not able to recover by a failure (I simulated this by changing perms to PGDATA folder) and see after how much time the "leadership" goes to replica. The result is that after doing this, patroni is continuously re-trying to start postgresql but with no luck, and replica node does not turn to leader even though it gets a response from leader that state is unknown.

I also want to point out that failover when messing with patroni (stop, restart) is responding as expected (replica becomes leader after a small amount of time and write access to db is restored to the new read-write node).

Leader logs
```
 2021-01-29 12:45:35,002 ERROR: Unexpected exception
 Traceback (most recent call last):
   File "/usr/local/lib/python3.6/site-packages/patroni/ha.py", line 1452, in run_cycle
     info = self._run_cycle()
   File "/usr/local/lib/python3.6/site-packages/patroni/ha.py", line 1356, in _run_cycle
     if self.state_handler.data_directory_empty():
   File "/usr/local/lib/python3.6/site-packages/patroni/postgresql/__init__.py", line 286, in data_directory_empty
     return data_directory_is_empty(self._data_dir)
   File "/usr/local/lib/python3.6/site-packages/patroni/utils.py", line 479, in data_directory_is_empty
     return all(os.name != 'nt' and (n.startswith('.') or n == 'lost+found') for n in os.listdir(data_dir))
 PermissionError: [Errno 13] Permission denied: '/var/lib/pgsql/12/data'
 2021-01-29 12:45:35,002 INFO: Unexpected exception raised, please report it as a BUG
```

Replica Logs
```
 2021-01-29 12:50:36,952 INFO: Got response from tsdb1 http://192.168.255.125:8008/patroni: {"state": "unknown", "role": "master", "database_system_identifier": "6919798977095709312", "patroni": {"version": "2.0.1", "scope": "clearview"}}
 2021-01-29 12:50:37,038 WARNING: Master (tsdb1) is still alive
 2021-01-29 12:50:37,076 INFO: following a different leader because i am not the healthiest node
```

Most of the configuration is copied from the templates (default values).

Configuration (same on both hosts, except local IPs of service)
```
scope: clearview
namespace: /service/
name: tsdb1

restapi:
  listen: xxx.125:8008
  connect_address: xxx.125:8008

etcd:
  host: xxx.127:2379

bootstrap:
  dcs:
    ttl: 30
    loop_wait: 10
    retry_timeout: 10
    maximum_lag_on_failover: 1048576
    postgresql:
      use_pg_rewind: true
      parameters:

  initdb:  # Note: It needs to be a list (some options need values, others are switches)
  - encoding: UTF8
  - data-checksums

  pg_hba:  # Add following lines to pg_hba.conf after running 'initdb'
  - host replication replicator xxx.125/32 md5
  - host replication replicator xxx.126/32 md5
  - host all all 0.0.0.0/0 md5

  users:
    admin:
      password: admin
      options:
        - createrole
        - createdb

postgresql:
  listen: xxx.125:5432
  connect_address: xxx.125:5432
  data_dir: /var/lib/pgsql/12/data
  bin_dir: /usr/pgsql-12/bin
  pgpass: /tmp/pgpass0
  pg_hba:  # Add following lines to pg_hba.conf after running 'initdb'
  - host replication replicator xxx.125/32 md5
  - host replication replicator xxx.126/32 md5
  - host all all 0.0.0.0/0 md5
  authentication:
    replication:
      username: replicator
      password:
    superuser:
      username: postgres
      password:
    rewind:  # Has no effect on postgres 10 and lower
      username: rewind_user
      password: 
  parameters:
    shared_preload_libraries: timescaledb
    unix_socket_directories: '.'

tags:
    nofailover: false
    noloadbalance: false
    clonefrom: false
    nosync: false
```

So basically my question is if this is the expected behavior, or, as I understood, when master_start_timeout reached, patroni should trigger the failover process, which is not happening now.

Thanks in advance

**Comment by CyberDem0n** [comment]
When you changed permissions of PGDATA you get system to the state when Patroni sort of lost control on PostgreSQL.
The last known state was that postgres running as primary. The current state could not really be determined, but Patroni running on the former leader still remembers the last state and therefore reports the role=master in the API. The key moment here is that the state could not be determined. What if Postgres is still running and serving writes? If we promote one of the replicas there is not guaranty that we will not get into the split-brain situation.

Anyway, what is the point of such exercise? What are you trying to test by changing permissions on PGDATA?

The master_start_timeout didn't really work here because according to logs Patroni didn't even try to start it up. It failed to determine the state of the node.

**Comment by thanasis-liapis** [comment]
The point of this exercise was to simulate a case where postgresql is inaccessible for clients for any reason that might be (indeed, I was not able to access the data anymore, nor write new), yet patroni is working and communicating with etcd. I wanted to see if this case is identified as a failure and triggers a switchover. The result in my test was to have two replicas (leader was turned to replica according to patronictl list of members) and no leader.

However, your explanation sure makes sense.

Thank you very much for your help.
